src/Microsoft.Security.Utilities.Core/HighPerformanceScanner.cs

```diff
@@ -0,0 +1,505 @@
+
+namespace Microsoft.Security.Utilities;
+
+#if NET8_0_OR_GREATER
+using System.Buffers;
+using StringInput = System.ReadOnlySpan<char>;
+#else
+using StringInput = string;
+#endif
+
+/// <summary>
+/// A detection from <see cref="HighPerformanceScanner"/>.
+/// </summary>
+internal readonly record struct HighPerformanceDetection
+{
+    /// <summary>
+    /// The signature that identifies the pattern detected.
+    /// </summary>
+    public string Signature { get; }
+
+    /// <summary>
+    /// The start index in the input where the pattern was detected.
+    /// </summary>
+    public int Start { get; }
+
+    /// <summary>
+    /// The length of the  in the input that matched in the
+    /// </summary>
+    public int Length { get; }
+
+    public HighPerformanceDetection(string signature, int start, int length)
+    {
+        Signature = signature;
+        Start = start;
+        Length = length;
+    }
+}
+
+/// <summary>
+/// A pattern that <see cref="HighPerformanceScanner"/> can scan for. These are
+/// typically the precomputed set derived from the patterns that implement <see
+/// cref="IHighPerformanceScannableKey"/>.
+/// </summary>
+internal sealed partial class CompiledHighPerformancePattern
+{
+    /// <summary>
+    /// The signature that identifies this pattern. This is searched for as fast
+    /// as possible in the input before applying the regex.
+    /// </summary>
+    public string Signature { get; }
+
+    /// <summary>
+    /// The fixed number of characters that appear before the signature in a match.
+    /// </summary>
+    public int PreSignatureLength { get; }
+
+    /// <summary>
+    /// The minimum number of characters that can appear in a match.
+    /// </summary>
+    public int MinLength { get; }
+
+    /// <summary>
+    /// The maximum number of characters that can appear in a match.
+    /// </summary>
+    public int MaxLength { get; }
+
+    /// <summary>
+    /// The regular expression that matches the pattern near a signature.
+    /// <summary>
+    /// <remarks>
+    /// WARNING: This is distinct from the general regexes for this pattern that
+    /// can be applied to find all matches in an arbitrary input. This regex is
+    /// not run on the entire input but only on a substring from <see
+    /// cref="PreSignatureLength" /> before a signature to <see cref="MaxLength"
+    /// /> from there."
+    ///
+    /// The whole expression must correspond to the secret. "refine" capture
+    /// group is not used.
+    ///
+    /// The regex msut also be anchored to the beginning of the input.
+    ///
+    /// The regex may assume that the signature is present at the expected
+    /// location. This can allow regexes to be shared that would differ only by
+    /// signature.
+    /// </remarks>
+    public Regex Regex { get; }
+
+    /// <summary>
+    /// A rendering of the signature for fast, allocation-free lookup. It is 4
+    /// ascii characters packed into a 32-bit integer.
```

Comment:
>packed into a 32-bit integer [](http://example.com/codeflow?start=25&length=28)

do we need a note on endianness here?

Reply:
I will note the byte order, but also note that it is an implementation detail.

Reply:
oh, I see, there's no unpacking of this thing, so who cares about the endianness, if your point. i agree it could be left out of a comment.

Reply:
Yes, exactly.